File: app/src/main/java/fi/tuska/jalkametri/activity/JalkametriActivity.java
/**
 * Copyright 2006-2011 Tuukka Haapasalo
 *
 * This file is part of jAlkaMetri.
 *
 * jAlkaMetri is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option)
 * any later version.
 *
 * jAlkaMetri is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with jAlkaMetri (LICENSE.txt). If not, see <http://www.gnu.org/licenses/>.
 */
package fi.tuska.jalkametri.activity;

import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences.Editor;
import android.content.res.Resources;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.Toast;
import fi.tuska.jalkametri.CommonActivities;
import fi.tuska.jalkametri.PrivateData;
import fi.tuska.jalkametri.R;
import fi.tuska.jalkametri.dao.Preferences;
import fi.tuska.jalkametri.data.PreferencesImpl;
import fi.tuska.jalkametri.util.BillingUtil;
import fi.tuska.jalkametri.util.TimeUtil;

/**
 * Abstract base class for jAlkaMetri activities; contains common
 * functionality.
 *
 * @author Tuukka Haapasalo
 */
public abstract class JalkametriActivity extends Activity implements GUIActivity {

    public static final int NO_HELP_TEXT = 0;

    private View mainView;
    private final int titleResourceId;
    private final int helpTextId;
    protected Preferences prefs;
    private boolean showDefaultHelpMenu = false;

    protected TimeUtil timeUtil;

    private boolean firstRun;

    /*
     * Standard activity functions
     * --------------------------------------------
     */
    protected JalkametriActivity(int titleResourceId, int helpTextId) {
        super();
        this.titleResourceId = titleResourceId;
        this.helpTextId = helpTextId;
    }

    /** Called when the activity is first created. */
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        this.timeUtil = new TimeUtil(this);

        this.prefs = new PreferencesImpl(this);

        Resources res = getResources();
        // Update title to enforce correct language
        setTitle(res.getString(titleResourceId));

        this.firstRun = true;
    }

    protected void findMainView() {
        this.mainView = findViewById(R.id.main_view);
    }

    /*
     * Custom actions
     * ----------------------------------------------------------
     */
    protected final void invalidateView() {
        if (mainView != null)
            mainView.invalidate();
    }

    protected void setShowDefaultHelpMenu(boolean state) {
        this.showDefaultHelpMenu = state;
    }

    @Override
    public Context getContext() {
        return getApplicationContext();
    }

    /**
     * Returns true on the first run, false afterwards. The value is reset at
     * onCreate(), so if this is called from onStart(), it will return true at
     * the first onStart() invocation, and false after that.
     */
    protected boolean isFirstRunAfterCreate() {
        boolean res = firstRun;
        firstRun = false;
        return res;
    }

    /**
     * Tries to suppress the soft keyboard from popping up automatically for
     * the given view.
     */
    protected void tryToHideSoftKeyboard(View view) {
        InputMethodManager service = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
        if (service != null) {
            service.hideSoftInputFromInputMethod(view.getWindowToken(),
                InputMethodManager.HIDE_IMPLICIT_ONLY);
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        boolean s = super.onCreateOptionsMenu(menu);
        if (showDefaultHelpMenu && helpTextId != NO_HELP_TEXT) {
            MenuInflater inflater = getMenuInflater();
            inflater.inflate(R.menu.menu_only_help, menu);
            return true;
        }
        return s;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (super.onOptionsItemSelected(item))
            return true;

        if (item.getItemId() == R.id.action_help) {
            // Show the help screen for this activity
            showHelp(null);
            return true;
        }

        return false;
    }

    @Override
    public boolean onContextItemSelected(MenuItem item) {
        switch (item.getItemId()) {

        case R.id.action_devel_clear_license:
            // Debugging action: clear license information
            if (PrivateData.DEVELOPER_FUNCTIONALITY_ENABLED) {
                BillingUtil.removeLicenseInfo(prefs);
                updateUI();
                Toast.makeText(this, "Licence cleared", Toast.LENGTH_SHORT).show();
                return true;
            }

        case R.id.action_devel_set_license:
            // Debugging action: set license information
            if (PrivateData.DEVELOPER_FUNCTIONALITY_ENABLED) {
                BillingUtil.setLicensePurchased(prefs);
                updateUI();
                Toast.makeText(this, "Licence set", Toast.LENGTH_SHORT).show();
                return true;
            }

        case R.id.action_devel_clear_def_library_created:
            // Debugging action: clear default drink library creation flag
            if (PrivateData.DEVELOPER_FUNCTIONALITY_ENABLED) {
                Editor editor = prefs.edit();
                prefs.setDrinkLibraryInitialized(editor, false);
                editor.commit();
                Toast.makeText(this, "Drink library flag cleared", Toast.LENGTH_SHORT).show();
                return true;
            }

        case R.id.action_devel_clear_disclaimer_read:
            // Debugging action: set license information
            if (PrivateData.DEVELOPER_FUNCTIONALITY_ENABLED) {
                Editor editor = prefs.edit();
                prefs.setDisclaimerRead(editor, false);
                editor.commit();
                Toast.makeText(this, "Disclaimer read flag reset", Toast.LENGTH_SHORT).show();
                return true;
            }
        }
        return super.onContextItemSelected(item);
    }

    /*
     * Common activity bindings --------------------------------------------
     */

    /**
     * Show the drink calculator.
     */
    public void showDrinkCalculator(View v) {
        CommonActivities.showCalculator(this);
    }

    /** Show the drinking history. */
    public void showDrinkHistory(View v) {
        CommonActivities.showDrinkHistory(this, prefs);
    }

    /** Shows the user preferences screen. */
    public void showPreferences(View v) {
        CommonActivities.showPreferences(this);
    }

    /** Shows the statistics screen. */
    public void showStatistics(View v) {
        CommonActivities.showStatistics(this);
    }

    /** Shows the statistics screen. */
    public void showHelp(View v) {
        if (helpTextId != NO_HELP_TEXT) {
            HelpActivity.showHelp(helpTextId, this);
        }
    }

}

File: app/src/main/java/fi/tuska/jalkametri/activity/JalkametriActivity.kt
package fi.tuska.jalkametri.activity

import android.app.Activity
import android.app.DialogFragment
import android.content.Context
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_HIDDEN
import android.view.inputmethod.InputMethodManager
import android.widget.Toast
import fi.tuska.jalkametri.Common.DEVELOPER_FUNCTIONALITY_ENABLED
import fi.tuska.jalkametri.CommonActivities
import fi.tuska.jalkametri.JalkametriApplication
import fi.tuska.jalkametri.R
import fi.tuska.jalkametri.dao.Preferences
import fi.tuska.jalkametri.util.LocaleHelper
import fi.tuska.jalkametri.util.TimeUtil
import java.util.Locale

/**
 * Abstract base class for jAlkaMetri activities; contains common
 * functionality.
 */
abstract class JalkametriActivity(private val titleResourceId: Int, private val helpTextId: Int) : Activity(), GUIActivity {

    private var mainView: View? = null
    private var showDefaultHelpMenu = false

    val jalkametriApplication: JalkametriApplication
        get() = application as JalkametriApplication

    val prefs: Preferences
        get() = jalkametriApplication.prefs

    val timeUtil: TimeUtil
        get() = jalkametriApplication.timeUtil

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val res = resources
        // Update title to enforce correct language
        title = res.getString(titleResourceId)
    }

    override fun setContentView(view: View) {
        super.setContentView(view)
        this.mainView = findViewById(R.id.main_view)
    }

    override fun setContentView(layoutResID: Int) {
        super.setContentView(layoutResID)
        this.mainView = findViewById(R.id.main_view)
    }

    override fun attachBaseContext(newBase: Context) {
        super.attachBaseContext(LocaleHelper.onAttach(newBase))
    }

    protected fun invalidateView() {
        mainView?.invalidate()
    }

    protected fun setShowDefaultHelpMenu(state: Boolean) {
        this.showDefaultHelpMenu = state
    }

    override fun getContext(): Context {
        return applicationContext
    }

    protected fun hideSoftKeyboard() {
        window.setSoftInputMode(SOFT_INPUT_STATE_ALWAYS_HIDDEN)
    }

    protected fun tryToHideSoftKeyboard(view: View) {
        val service = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        service.hideSoftInputFromInputMethod(view.windowToken, 0)
    }

    protected fun showCustomDialog(dialog: DialogFragment) = showCustomDialog(this, dialog)

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val s = super.onCreateOptionsMenu(menu)
        if (showDefaultHelpMenu && helpTextId != NO_HELP_TEXT) {
            val inflater = menuInflater
            inflater.inflate(R.menu.menu_only_help, menu)
            return true
        }
        return s
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (super.onOptionsItemSelected(item))
            return true

        if (item.itemId == R.id.action_help) {
            // Show the help screen for this activity
            showHelp(null)
            return true
        }

        return false
    }

    override fun onContextItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {

            R.id.action_devel_clear_def_library_created ->
                // Debugging action: clear default drink library creation flag
                if (DEVELOPER_FUNCTIONALITY_ENABLED) {
                    prefs.edit().let {
                        prefs.setDrinkLibraryInitialized(it, false)
                        it.commit()
                    }
                    Toast.makeText(this, "Drink library flag cleared", Toast.LENGTH_SHORT).show()
                    return true
                }

            R.id.action_devel_clear_disclaimer_read ->
                // Debugging action: set license information
                if (DEVELOPER_FUNCTIONALITY_ENABLED) {
                    prefs.edit().let {
                        prefs.setDisclaimerRead(it, false)
                        it.commit()
                    }
                    Toast.makeText(this, "Disclaimer read flag reset", Toast.LENGTH_SHORT).show()
                    return true
                }
        }
        return super.onContextItemSelected(item)
    }

    open fun showDrinkCalculator(v: View?) = CommonActivities.showCalculator(this)
    fun showDrinkHistory(v: View?) = CommonActivities.showDrinkHistory(this, prefs)
    fun showPreferences(v: View?) = CommonActivities.showPreferences(this)
    fun showStatistics(v: View?) = CommonActivities.showStatistics(this)

    fun showHelp(v: View?) {
        if (helpTextId != NO_HELP_TEXT) {
            HelpActivity.showHelp(helpTextId, this)
        }
    }

    companion object {
        val NO_HELP_TEXT = 0

        fun showCustomDialog(activity: Activity, dialog: DialogFragment) {
            // DialogFragment.show() will take care of adding the fragment
            // in a transaction.  We also want to remove any currently showing
            // dialog, so make our own transaction and take care of that here.
            val fm = activity.fragmentManager
            val ft = fm.beginTransaction()
            val prev = fm.findFragmentByTag("dialog")
            if (prev != null) {
                ft.remove(prev)
            }
            ft.addToBackStack(null)
            dialog.show(ft, "dialog")
        }

    }

}
